### ISSUE TITLE
doc_overindented_list_items false positive expecting fewer spaces

### ISSUE THREAD

**Issue by azriel91** [rust-lang/rust-clippy#14275]
### Summary

Heya, when a markdown list item has an inner block of text, I think the correct markdown convention is to indent the inner block by 4 spaces relative to the list item.

When the inner block of text spans at least two lines, then clippy now warns on `doc_overindented_list_items`.

This used to not err on nightly from maybe 2 months ago; I just updated today (2025-02-22).

```bash
cargo clippy -- --version
clippy 0.1.86 (f04bbc60f8 2025-02-20)

# I think the commit hash is rustc's commit hash,
# but I think the version is correct

rustc --version
rustc 1.87.0-nightly (f04bbc60f 2025-02-20)
```

### Lint Name

doc_overindented_list_items

### Reproducer

I tried this code:

```rust
//! This is a list:
//!
//! 1. Item one which has a long long long long text even longer onto the next
//!    line.
//!
//!     Some details which has a long long long long text even longer onto the
//!     next line.

fn main() {
    println!("Hello, world!");
}
```

When running `cargo clippy -- -D warnings`, I saw this happen:

```
    Checking doc_overindent v0.1.0 (C:\Users\azrielh\AppData\Local\Temp\doc_overindent)
error: doc list item overindented
 --> src\main.rs:7:5
  |
7 | //!     next line.
  |     ^^^^ help: try using `   ` (3 spaces)
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#doc_overindented_list_items
  = note: `-D clippy::doc-overindented-list-items` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(clippy::doc_overindented_list_items)]`

error: could not compile `doc_overindent` (bin "doc_overindent") due to 1 previous error
```

I expected to see this happen:

### Version

```text

```

**Comment by azriel91** [comment]
heya @ohno418, just noticed you implemented this lint in 4693d0a, would you be free to extend the lint to handle this case?

I like it for most of my docs, except sometimes there's a lot of subtext beneath the list item 😅.

**Comment by ohno418** [comment]
Hmm..., I'm not sure how we should handle a paragraph that is separated from a list item by a blank line. I mean, I wonder if it should  really be considered part of the list item or not. If not, Clippy shouldn't warn about the paragraph.

I'll take a closer look into this when I have time. Thanks for letting me know, anyway.

**Comment by ohno418** [comment]
(And I'm not sure why Clippy doesn't warn about the "Some details ..." line, while it does warn the "next line." line...)

**Comment by MichaReiser** [comment]
Thanks for implementing this rule. It caught a few true-positives in our code base. 

However, I think there's one other case where this rule raises an issue when I think it shouldn't and seems related:

```
/// 2. Else, evaluate `fallback_fn()`:
///    a. If `self` is definitely unbound, return the result of `fallback_fn()`.
///    b. Else, if `fallback` is definitely unbound, return `self`.
///    c. Else, if `self` is possibly unbound and `fallback` is definitely bound,
///       return `Symbol(<union of self-type and fallback-type>, Boundness::Bound)`
///    d. Else, if `self` is possibly unbound and `fallback` is possibly unbound,
///       return `Symbol(<union of self-type and fallback-type>, Boundness::PossiblyUnbound)`
```


```
    Checking ruff_python_resolver v0.0.0 (/Users/micha/astral/ruff/crates/ruff_python_resolver)
warning: doc list item overindented
   --> crates/red_knot_python_semantic/src/symbol.rs:470:9
    |
470 |     ///      return `Symbol(<union of self-type and fallback-type>, Boundness::Bound)`
    |         ^^^^^ help: try using `   ` (3 spaces)
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#doc_overindented_list_items
    = note: `#[warn(clippy::doc_overindented_list_items)]` on by default

warning: doc list item overindented
   --> crates/red_knot_python_semantic/src/symbol.rs:472:9
    |
472 |     ///      return `Symbol(<union of self-type and fallback-type>, Boundness::PossiblyUnbound)`
    |         ^^^^^ help: try using `   ` (3 spaces)
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#doc_overindented_list_items
```

Let me know if you prefer me opening a new issue

**Comment by ohno418** [comment]
Hi @MichaReiser, thank you for the report.

I had a quick look into this and found that rustdoc doesn't seem to support lists written like:

```rust
/// a. first item.
/// b. next item.
```

You can see this when generating documentation with rustdoc.
(CommonMark, the Markdown specification that rustdoc uses, doesn't support this style either.)

As a result, rustdoc and clippy interpret it as normal text, which leads to the output you saw.

If you want to write a nested ordered list, using numbers would work.

**Comment by MichaReiser** [comment]
Oh, I wasn't aware that rustdoc doesn't support lists written with `a)`, `b)`... Sorry for the noise!

**Comment by andrei-pashkin** [comment]
Even simpler example - a bullet-point list where one item has multiple lines:
```rust
/// Greets someone.
/// 
/// # Arguments
/// 
/// - `arg`: Someone you want 
///          to greet.
pub fn foo(arg: &str) {
    println!("Hello, {}!", arg);
}
```

Clippy gives `warning: doc list item overindented` to this.

**Comment by ohno418** [comment]
@AndreiPashkin 
Hmm, I don't think this is a false positive. rustdoc and Clippy don't seem to treat `xxx: aaaaa` as structured text; they just treat it as normal text.

**Comment by AndreiPashkin** [comment]
@ohno418, hmm, maybe I'm missing something, but I didn't get the error on such lists until very recently.

**Comment by Amberley-Sz** [comment]
Hi, it seems like the overindentation check conflicts with the lazy_continuation check.
I got 25 overindentation check errors like this:
```
   |
28 | //!    [`Instrument`] and [`WithCollector`] implementations for
   |     ^^^ help: try using `  ` (2 spaces)
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#doc_overindented_list_items
   = note: `-D clippy::doc-overindented-list-items` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::doc_overindented_list_items)]`
```
When I tried use 2 spaces as it suggested, I got the doc_lazy_continuation check error saying that I don't have indentation on this same item
```
error: doc list item without indentation
  --> tracing-futures/src/lib.rs:28:5
   |
28 | //!  [`Instrument`] and [`WithCollector`] implementations for
   |     ^
   |
   = help: if this is supposed to be its own paragraph, add a blank line
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#doc_lazy_continuation
   = note: `-D clippy::doc-lazy-continuation` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::doc_lazy_continuation)]`
help: indent this line
   |
28 | //!   [`Instrument`] and [`WithCollector`] implementations for
   |  
```
Is this a bug?

**Comment by ohno418** [comment]
@Amberley-Sz 
Based on the suggestions you received, it seems that you should to put 3 spaces
between `//!` and `[`Instrument`]`. Two of the suggestions said exactly this.
(Please be careful about what the first suggestion says. It says to use 2 spaces
between `//! ` (with a single space after `//!`) and `[`Instrument`]`.)

So, try this:

```
28 | //!   [`Instrument`] and [`WithCollector`] implementations for
```

**Comment by Amberley-Sz** [comment]
@ohno418 Thank you so much for the clarification! It works now!

**Comment by jyn514** [comment]
> Hmm, I don't think this is a false positive. rustdoc and Clippy don't seem to treat `xxx: aaaaa` as structured text; they just treat it as normal text.

this is not structured text. this is a markdown list; you can see the `-` at the front. for example this code:
```
    /// - `YDB_ERR_TIME2LONG` if `timeout.as_nanos()` exceeds `YDB_MAX_TIME_NSEC`
    ///                    or if `timeout.as_nanos()` does not fit into a `c_ulonglong`.
```
renders correctly, but gives a clippy warning:
![Image](https://github.com/user-attachments/assets/c483c37f-2162-4c77-8f9d-62ff9b247e4f)
```
warning: doc list item overindented
    --> src/context_api/mod.rs:1047:9
     |
1047 |     ///                    or if `timeout.as_nanos()` does not fit into a `c_ulonglong`.
     |         ^^^^^^^^^^^^^^^^^^^ help: try using `  ` (2 spaces)
     |
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#doc_overindented_list_items
```

**Comment by jyn514** [comment]
turning it down to allow by default and backporting that would be good, yes.

i think it’s possible to make a version of this lint with sufficiently few false positives it can be on by default. but i don’t think the current implementation is that version.

**Comment by pitaj** [comment]
> this is structured text. this is a markdown list; you can see the - at the front. for example this code:
```rust
    /// - `YDB_ERR_TIME2LONG` if `timeout.as_nanos()` exceeds `YDB_MAX_TIME_NSEC`
    ///                    or if `timeout.as_nanos()` does not fit into a `c_ulonglong`.
```

This is not a false positive IMO. Structuring the text with the extra spaces before "or if..." is like aligning assignments:
```rust
let foo_height = 45;
let foo_width  = 12;
```
It can improve readability in some cases, but we don't support it in rustfmt.

Also, it can be misleading. For instance, one writing that documentation may expect it to be displayed in the same way, with the line break. But as your image clearly shows, that is not the case.

**Comment by pitaj** [comment]
> this is structured text. this is a markdown list; you can see the - at the front. for example this code:
```rust
    /// - `YDB_ERR_TIME2LONG` if `timeout.as_nanos()` exceeds `YDB_MAX_TIME_NSEC`
    ///                    or if `timeout.as_nanos()` does not fit into a `c_ulonglong`.
```

This is not a false positive IMO. Structuring the text with the extra spaces before "or if..." is like aligning assignments:
```rust
let foo_height = 45;
let foo_width  = 12;
```
It can improve readability in some cases, but we don't support it in rustfmt.

Also, it can be misleading. For instance, one writing that documentation may expect it to be displayed in the same way, with the line break. But as your image clearly shows, that is not the case.

The same applies to @AndreiPashkin 's code as well: 

```rust
/// Greets someone.
/// 
/// # Arguments
/// 
/// - `arg`: Someone you want 
///          to greet.
```

While it's kinda nice to have the argument description in a separate column from the argument name, markdown has no such feature.

**Comment by AndreiPashkin** [comment]
@pitaj,

> While it's kinda nice to have the argument description in a separate column from the argument name, markdown has no such feature.

Markdown has no such feature, but it also doesn't prohibit such formatting:
```
$ pandoc --from markdown_strict --to html ~/test.md 
<h1>Arguments</h1>
<ul>
<li><code>arg</code>: Someone you want to greet.</li>
</ul>
```

Such formatting is a standard Markdown.

**Comment by LesnyRumcajs** [comment]
This also fails
```
/// 1) root_d (32 bytes)
/// 2) root_c (32 bytes)
/// 3) root_r (32 bytes)
/// 4) For each synthetic challenge proof:
///    4.1) Challenge's node index (8 bytes)
///    4.2) Parents' node indices (8 bytes per parent)
///    4.3) Challenge's proof_d (32 bytes for leaf_d and 32 bytes per path_d sibling)
///    4.4) Challenge's column (32 bytes per layer)
///    4.5) Challenge's proof_c (32 bytes for leaf_c and 32 bytes per path_c sibling)
///    4.6) For each parent:
///       4.6.1) Parent's column (32 bytes per layer)
///       4.6.2) Parent's proof_c (32 bytes for leaf_c and 32 bytes per path_c sibling)
///    4.7) Challenge's proof_r (32 bytes for leaf_r and 32 bytes per path_r sibling)
```

```
   --> storage-proofs-porep/src/stacked/vanilla/params.rs:481:5
    |
481 | ///        4.6.2) Parent's proof_c (32 bytes for leaf_c and 32 bytes per path_c sibling)
    |     ^^^^^^^ help: try using `   ` (3 spaces)
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.93.0/index.html#doc_overindented_list_items
```

**Comment by ohno418** [comment]
@LesnyRumcajs 
This seems to be because the list doesn't follow CommonMark's list formatting rule. You can verify this by running `cargo doc` . The output won't render it as a list.

**Comment by LesnyRumcajs** [comment]
> [@LesnyRumcajs](https://github.com/LesnyRumcajs) This seems to be because the list doesn't follow CommonMark's list formatting rule. You can verify this by running `cargo doc` . The output won't render it as a list.

Indeed, you're right! Thanks for clarifying.
